pkg/appgw/backendaddresspools.go

```diff
@@ -7,82 +7,121 @@ package appgw
-					for _, address := range subset.Addresses {
-						ip := address.IP
-						hostname := address.Hostname
-						// prefer IP address
-						if len(ip) != 0 {
-							// address specified by ip
-							addressPoolAddresses = append(addressPoolAddresses, network.ApplicationGatewayBackendAddress{IPAddress: &ip})
-						} else if len(address.Hostname) != 0 {
-							// address specified by hostname
-							addressPoolAddresses = append(addressPoolAddresses, network.ApplicationGatewayBackendAddress{Fqdn: &hostname})
-						}
-					}
+func getBackendAddressMapKeys(m *map[n.ApplicationGatewayBackendAddress]interface{}) *[]n.ApplicationGatewayBackendAddress {
+	var addresses []n.ApplicationGatewayBackendAddress
+	for addr := range *m {
+		addresses = append(addresses, addr)
+	}
+	sort.Sort(byIPFQDN(addresses))
+	return &addresses
+}
 
-					addressPool = network.ApplicationGatewayBackendAddressPool{
-						Etag: to.StringPtr("*"),
-						Name: &addressPoolName,
-						ApplicationGatewayBackendAddressPoolPropertiesFormat: &network.ApplicationGatewayBackendAddressPoolPropertiesFormat{
-							BackendAddresses: &addressPoolAddresses,
-						},
-					}
+func getPorts(subset v1.EndpointSubset) map[int32]interface{} {
+	ports := make(map[int32]interface{})
+	for _, endpointsPort := range subset.Ports {
+		if endpointsPort.Protocol == v1.ProtocolTCP {
+			ports[endpointsPort.Port] = nil
+		}
+	}
+	return ports
+}
 
-					addressPools[*addressPool.Name] = addressPool
-				}
+func (builder *appGwConfigBuilder) getBackendAddressPool(backendID backendIdentifier, serviceBackendPair serviceBackendPortPair, defaultPool *n.ApplicationGatewayBackendAddressPool, addressPools *map[string]*n.ApplicationGatewayBackendAddressPool) *n.ApplicationGatewayBackendAddressPool {
+	endpoints := builder.k8sContext.GetEndpointsByService(backendID.serviceKey())
+	if endpoints == nil {
+		logLine := fmt.Sprintf("Unable to get endpoints for service key [%s]", backendID.serviceKey())
+		builder.recorder.Event(backendID.Ingress, v1.EventTypeWarning, "EndpointsEmpty", logLine)
+		glog.Warning(logLine)
+		return defaultPool
+	}
 
-				builder.backendPoolMap[backendID] = &addressPool
-				break
+	for _, subset := range endpoints.Subsets {
+		if _, portExists := getPorts(subset)[serviceBackendPair.BackendPort]; portExists {
+			poolName := generateAddressPoolName(backendID.serviceFullName(), backendID.Backend.ServicePort.String(), serviceBackendPair.BackendPort)
+			// The same service might be referenced in multiple ingress resources, this might result in multiple `serviceBackendPairMap` having the same service key but different
+			// ingress resource. Thus, while generating the backend address pool, we should make sure that we are generating unique backend address pools.
+			pool, ok := (*addressPools)[poolName]
+			if !ok {
+				pool = newPool(poolName, subset)
+				(*addressPools)[poolName] = pool
 			}
+			return pool
 		}
```

Comment:
If we are not able to find the `serviceBackendPair.BackendPort`, should we raise an event that some subnet in endpoint don't have the referenced `port` ?